src/main/java/com/direwolf20/buildinggadgets/common/util/blocks/RegionSnapshot.java

```diff
@@ -0,0 +1,247 @@
+package com.direwolf20.buildinggadgets.common.util.blocks;
+
+import com.direwolf20.buildinggadgets.api.building.Region;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.nbt.NBTUtil;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import net.minecraft.world.dimension.DimensionType;
+import net.minecraftforge.common.util.BlockSnapshot;
+import net.minecraftforge.common.util.Constants;
+import net.minecraftforge.fml.server.ServerLifecycleHooks;
+
+import java.util.*;
+import java.util.function.BiPredicate;
+
+public class RegionSnapshot {
+
+    private static final String DIMENSION = "dim";
+    private static final String BLOCK_FRAMES = "frames";
+    private static final String BLOCK_PALETTES = "palettes";
+    private static final String BLOCK_SNAPSHOTS = "block_snapshots";
+
+    private static NBTTagCompound serialize(NBTTagCompound tag, RegionSnapshot snapshot) {
+        tag.setInt(DIMENSION, snapshot.world.getDimension().getType().getId());
+        snapshot.region.serializeTo(tag);
+
+        // Palette serialization begin
+        NBTTagList palettes = new NBTTagList();
+        Object2IntMap<IBlockState> mapPalettes = new Object2IntOpenHashMap<>();
+        Set<IBlockState> recordedPalettes = new HashSet<>();
+
+        // "nothing should be here" block state
+        palettes.add(new NBTTagCompound());
+        // Palette serialization end
+
+        // Streaking: a way to record a number of repeating block states, using only 2 integers
+        // Storage frame: one or two integers indicating a sequence of block states
+        // Regular frame: a type of storage frame; 1 integer, representing the ID of the block state recorded
+        // Streak frame: a type of storage frame; 2 integers, where the first one is positive, representing the ID of the repeating block state; the second one is negative, representing the number of repeating block states
```

Comment:
That was my mistake: things should be in `short`. Also is short is 2 bytes long, or 16 bits, within range of 32767 to -32768.

Reply:
Java primitive types: 
byte: 1 byte
short: 2 bytes
int: 4 bytes
long 8 bytes
All types are signed. 
How can something be 2 bytes long and not have 16bits?

Reply:
Also my first part of the comment was propably a bit misleading - as I revised it whilst writing the comment. What would be best in my opinion, would be to just have 1 byte streak indicator (0==no streak; treating it as if it would be signed) and then have 2-4 (presumably 3 in order to make life easier when reading (int instead of byte List)) bytes with the state id.
We should also treat the state id as of it would be unsigned in my opinion... 

I'm not sure if we want to use BlockSnapshot, if we are optimising on that level...

Reply:
You method should work better :+1:.  Additional point for you design if we are going to use this class for Moving Gadget.
Also `BlockSnapshot` is only used for tile entities, normal block states are stored with palettes.

Reply:
`BlockSnapshot` is only used for tile entities, pure blocks are stored using palettes.